R/create_codemeta.R

```diff
@@ -14,155 +14,198 @@
-                                              cm$identifier)))
-          }else{
-            if(provider$name == "BioConductor"){
-              cm$relatedLink <- unique(c(cm$relatedLink,
-                                         paste0("https://bioconductor.org/packages/release/bioc/html/",
-                                                cm$identifier, ".html")))
-            }
-          }
-        }
-      }else{
-        if(cm$identifier %in% installed_package_names()){
-          pkg_info <- sessioninfo::package_info(cm$identifier)
-          pkg_info <- pkg_info[pkg_info$package == cm$identifier,]
-          provider_name <- pkg_info$source
-          if(cm$version == pkg_info$ondiskversion){
-            if(grepl("CRAN", provider_name)){
-              cm$relatedLink <- unique(c(cm$relatedLink,
-                                         paste0("https://CRAN.R-project.org/package=",
-                                                cm$identifier)))
-            }else{
-              if(grepl("Bioconductor", provider_name)){
-                cm$relatedLink <- unique(c(cm$relatedLink,
-                                           paste0("https://bioconductor.org/packages/release/bioc/html/",
-                                                  cm$identifier, ".html")))
-              }else{
-                # if GitHub try to build the URL to commit or to repo in general
-                if(grepl("Github", provider_name)){
-                  if(grepl("@", provider_name)){
-                    commit <- gsub(".*@", "", provider_name)
-                    commit <- gsub("\\)", "", commit)
-                    link <- gsub(".*\\(", "", provider_name)
-                    link <- gsub("@.*", "", link)
-                    cm$relatedLink <- unique(c(cm$relatedLink,
-                                               paste0("https://github.com/", link,
-                                                      "/commit/", commit)))
-                  }
-                }
-              }
-            }
-          }
-        }
+  provider <- guess_provider(cm$identifier)
+
+  if (! is.null(provider)) {
+
+    readme <- guess_readme(root)$readme_path
+
+    if (! is.null(readme)) {
+
+      badges <- extract_badges(readme)
 
+      if (! is.null(provider) &&
+          whether_provider_badge(badges, provider$name)) {
+
+        cm <- set_relatedLink_1(cm, provider)
       }
-    }
+    } else if (cm$identifier %in% installed_package_names()) {
 
+      pkg_info <- sessioninfo::package_info(cm$identifier)
+      pkg_info <- pkg_info[pkg_info$package == cm$identifier, ]
+      provider_name <- pkg_info$source
 
+      if (cm$version == pkg_info$ondiskversion) {
 
+        cm <- set_relatedLink_2(cm, provider_name)
+      }
+    }
+  }
 
   ## Add blank slots as placeholders? and declare as an S3 class?
-
   cm
 }
+
+# set_relatedLink_1 ------------------------------------------------------------
+set_relatedLink_1 <- function(codemeta, provider) {
```

Comment:
could the two helpers have more informative names, even if longer?

Reply:
see below